File: src/main/java/org/jcodec/containers/mp4/demuxer/DashMP4DemuxerTrack.java
package org.jcodec.containers.mp4.demuxer;

import static org.jcodec.common.TrackType.AUDIO;
import static org.jcodec.common.TrackType.OTHER;
import static org.jcodec.common.TrackType.VIDEO;
import static org.jcodec.common.VideoCodecMeta.createSimpleVideoCodecMeta;

import java.io.Closeable;
import java.io.File;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

import org.jcodec.codecs.aac.AACUtils;
import org.jcodec.codecs.h264.H264Utils;
import org.jcodec.codecs.h264.io.model.SeqParameterSet;
import org.jcodec.codecs.h264.mp4.AvcCBox;
import org.jcodec.common.ArrayUtil;
import org.jcodec.common.AudioCodecMeta;
import org.jcodec.common.Codec;
import org.jcodec.common.DemuxerTrackMeta;
import org.jcodec.common.SeekableDemuxerTrack;
import org.jcodec.common.TrackType;
import org.jcodec.common.VideoCodecMeta;
import org.jcodec.common.io.NIOUtils;
import org.jcodec.common.io.SeekableByteChannel;
import org.jcodec.common.model.ColorSpace;
import org.jcodec.common.model.Packet.FrameType;
import org.jcodec.containers.mp4.MP4Packet;
import org.jcodec.containers.mp4.MP4TrackType;
import org.jcodec.containers.mp4.MP4Util;
import org.jcodec.containers.mp4.MP4Util.Atom;
import org.jcodec.containers.mp4.boxes.AudioSampleEntry;
import org.jcodec.containers.mp4.boxes.MovieBox;
import org.jcodec.containers.mp4.boxes.MovieFragmentBox;
import org.jcodec.containers.mp4.boxes.NodeBox;
import org.jcodec.containers.mp4.boxes.PixelAspectExt;
import org.jcodec.containers.mp4.boxes.SampleEntry;
import org.jcodec.containers.mp4.boxes.TrackFragmentBaseMediaDecodeTimeBox;
import org.jcodec.containers.mp4.boxes.TrackFragmentBox;
import org.jcodec.containers.mp4.boxes.TrakBox;
import org.jcodec.containers.mp4.boxes.TrunBox;
import org.jcodec.containers.mp4.boxes.VideoSampleEntry;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * Demuxes one track out of multiple DASH fragments
 * 
 * @author The JCodec project
 * 
 */
public class DashMP4DemuxerTrack implements SeekableDemuxerTrack, Closeable {

    private int[][] sizes;
    private int[][] compOffsets;
    private long[] chunkOffsets;
    private int[] avgDur;
    private int[][] sampleDurations;
    private long offInChunk;
    private SeekableByteChannel[] inputs;
    private int curFrame;
    private int curFrag;
    private long globalFrame;
    private long pts;
    private int frameCount;
    private long totalDuration;
    private TrakBox trak;
    private SampleEntry[] sampleEntries;
    private double durationHint;

    public static class Fragment {
        TrackFragmentBox frag;
        long offset;
        SeekableByteChannel input;

        public Fragment(TrackFragmentBox frag, long offset, SeekableByteChannel input) {
            this.frag = frag;
            this.offset = offset;
            this.input = input;
        }
    }

    public static class FragmentComparator implements Comparator<Fragment> {
        @Override
        public int compare(Fragment arg0, Fragment arg1) {
            long a = arg0.frag.getTfdt() == null ? 0 : arg0.frag.getTfdt().getBaseMediaDecodeTime();
            long b = arg1.frag.getTfdt() == null ? 0 : arg1.frag.getTfdt().getBaseMediaDecodeTime();
            return a < b ? -1 : (a == b ? 0 : 1);
        }
    }

    public DashMP4DemuxerTrack(MovieBox mov, TrakBox trak, Fragment[] fragments) {
        long prevOffset = 0;
        int prevSampleCount = 0;
        sizes = new int[fragments.length][];
        compOffsets = new int[fragments.length][];
        chunkOffsets = new long[fragments.length];
        avgDur = new int[fragments.length];
        sampleDurations = new int[fragments.length][];
        this.trak = trak;
        inputs = new SeekableByteChannel[fragments.length];
        sampleEntries = NodeBox.findAllPath(trak, SampleEntry.class,
                new String[] { "mdia", "minf", "stbl", "stsd", null });

        if (sortable(fragments))
            Arrays.sort(fragments, 0, fragments.length, new FragmentComparator());

        int i = 0;
        for (Fragment fragment : fragments) {
            TrackFragmentBox frag = fragment.frag;
            TrunBox trun = frag.getTrun();
            TrackFragmentBaseMediaDecodeTimeBox tfdt = frag.getTfdt();
            if (tfdt != null) {
                if (i > 0) {
                    avgDur[i - 1] = (int) ((tfdt.getBaseMediaDecodeTime() - prevOffset) / prevSampleCount);
                    totalDuration = tfdt.getBaseMediaDecodeTime();
                }
                prevOffset = tfdt.getBaseMediaDecodeTime();
            }
            prevSampleCount = (int) trun.getSampleCount();
            sizes[i] = trun.getSampleSizes();
            compOffsets[i] = trun.getSampleCompositionOffsets();
            chunkOffsets[i] = fragment.offset + frag.getTfhd().getBaseDataOffset() + frag.getTrun().getDataOffset();
            if (trun.isSampleDurationAvailable()) {
                sampleDurations[i] = trun.getSampleDurations();
                totalDuration += ArrayUtil.sumInt(sampleDurations[i]);
            }
            frameCount += sizes[i].length;
            inputs[i] = fragment.input;
            i++;
        }
        if (avgDur.length > 1) {
            avgDur[avgDur.length - 1] = avgDur[avgDur.length - 2];
            totalDuration += avgDur[avgDur.length - 1] * prevSampleCount;
        }
    }

    private boolean sortable(Fragment[] fragments) {
        for (Fragment fragment : fragments) {
            if (fragment.frag.getTfdt() == null)
                return false;
        }
        return true;
    }

    public static DashMP4DemuxerTrack createFromFiles(List<File> files) throws IOException {
        List<Fragment> fragments = new ArrayList<Fragment>();

        MovieBox moov = null;
        for (File file : files) {
            SeekableByteChannel channel = NIOUtils.readableChannel(file);
            for (Atom atom : MP4Util.getRootAtoms(channel)) {
                if ("moov".equals(atom.getHeader().getFourcc())) {
                    moov = (MovieBox) atom.parseBox(channel);
                } else if ("moof".equalsIgnoreCase(atom.getHeader().getFourcc())) {
                    MovieFragmentBox mfra = (MovieFragmentBox) atom.parseBox(channel);
                    TrackFragmentBox tfra = mfra.getTracks()[0];
                    fragments.add(new Fragment(tfra, atom.getOffset(), channel));
                }
            }

        }
        return new DashMP4DemuxerTrack(moov, moov.getTracks()[0], fragments.toArray(new Fragment[0]));
    }

    @Override
    public synchronized MP4Packet nextFrame() throws IOException {
        if (curFrag >= sizes.length)
            return null;
        if (curFrame >= sizes[curFrag].length) {
            curFrag++;
            curFrame = 0;
            offInChunk = 0;
        }
        if (curFrag >= sizes.length)
            return null;
        int size = sizes[curFrag][(int) curFrame];

        return getNextFrame(ByteBuffer.allocate(size));
    }

    protected ByteBuffer readPacketData(SeekableByteChannel input, ByteBuffer buffer, long offset, int size)
            throws IOException {
        ByteBuffer result = buffer.duplicate();
        synchronized (input) {
            input.setPosition(offset);
            NIOUtils.readL(input, result, size);
        }
        result.flip();
        return result;
    }

    public synchronized MP4Packet getNextFrame(ByteBuffer storage) throws IOException {
        if (curFrag >= sizes.length)
            return null;
        if (curFrame >= sizes[curFrag].length)
            return null;
        int size = sizes[curFrag][(int) curFrame];

        if (storage != null && storage.remaining() < size) {
            throw new IllegalArgumentException("Buffer size is not enough to fit a packet");
        }

        long pktPos = chunkOffsets[curFrag] + offInChunk;

        ByteBuffer result = readPacketData(inputs[curFrag], storage, pktPos, size);

        if (result != null && result.remaining() < size)
            return null;

        int duration = sampleDurations[curFrag] == null
                ? (avgDur[curFrag] == 0 ? (int) (durationHint * trak.getTimescale()) : avgDur[curFrag])
                : sampleDurations[curFrag][curFrame];

        boolean sync = curFrame == 0;

        long realPts = pts;
        if (compOffsets[curFrag] != null) {
            realPts = pts + compOffsets[curFrag][curFrame];
        }

        FrameType ftype = sync ? FrameType.KEY : FrameType.INTER;
        MP4Packet pkt = new MP4Packet(result, realPts, trak.getTimescale(), duration, globalFrame, ftype, null, 0,
                realPts, 1, pktPos, size, false);

        offInChunk += size;

        pts += duration;
        curFrame++;
        globalFrame++;

        return pkt;
    }

    @Override
    public DemuxerTrackMeta getMeta() {
        MP4TrackType type = TrakBox.getTrackType(trak);
        TrackType t = type == MP4TrackType.VIDEO ? VIDEO : (type == MP4TrackType.SOUND ? AUDIO : OTHER);
        int[] seekFrames = new int[sizes.length];
        for (int i = 0, numFrames = 0; i < sizes.length; i++) {
            seekFrames[i] = numFrames;
            numFrames += sizes[i].length;
        }

        ByteBuffer codecPrivate = getCodecPrivate();
        VideoCodecMeta videoCodecMeta = getVideoCodecMeta();
        AudioCodecMeta audioCodecMeta = getAudioCodecMeta();
        return new MP4DemuxerTrackMeta(t, Codec.codecByFourcc(sampleEntries[0].getFourcc()), totalDuration, seekFrames,
                frameCount, codecPrivate, videoCodecMeta, audioCodecMeta, sampleEntries);
    }

    protected ColorSpace getColorInfo() {
        Codec codec = Codec.codecByFourcc(trak.getFourcc());
        if (codec == Codec.H264) {
            AvcCBox avcC = H264Utils.parseAVCC((VideoSampleEntry) sampleEntries[0]);
            List<ByteBuffer> spsList = avcC.getSpsList();
            if (spsList.size() > 0) {
                SeqParameterSet sps = SeqParameterSet.read(spsList.get(0).duplicate());
                return sps.getChromaFormatIdc();
            }
        }
        return null;
    }

    private AudioCodecMeta getAudioCodecMeta() {
        MP4TrackType type = TrakBox.getTrackType(trak);
        AudioCodecMeta audioCodecMeta = null;
        if (type == MP4TrackType.SOUND) {
            AudioSampleEntry ase = (AudioSampleEntry) sampleEntries[0];
            audioCodecMeta = AudioCodecMeta.fromAudioFormat(ase.getFormat());
        }
        return audioCodecMeta;
    }

    private VideoCodecMeta getVideoCodecMeta() {
        MP4TrackType type = TrakBox.getTrackType(trak);
        VideoCodecMeta videoCodecMeta = null;
        if (type == MP4TrackType.VIDEO) {
            videoCodecMeta = createSimpleVideoCodecMeta(trak.getCodedSize(), getColorInfo());
            PixelAspectExt pasp = NodeBox.findFirst(sampleEntries[0], PixelAspectExt.class, "pasp");
            if (pasp != null)
                videoCodecMeta.setPixelAspectRatio(pasp.getRational());
        }
        return videoCodecMeta;
    }

    public ByteBuffer getCodecPrivate() {
        Codec codec = Codec.codecByFourcc(sampleEntries[0].getFourcc());
        if (codec == Codec.H264) {
            AvcCBox avcC = H264Utils.parseAVCC((VideoSampleEntry) sampleEntries[0]);
            return H264Utils.avcCToAnnexB(avcC);

        } else if (codec == Codec.AAC) {
            return AACUtils.getCodecPrivate(sampleEntries[0]);
        }
        // This codec does not have private section
        return null;
    }

    @Override
    public boolean gotoFrame(long frameNo) throws IOException {
        int curFrag = 0;
        int globalFrame = 0;
        int pts = 0;
        for (int[] is : sizes) {
            if (frameNo > is.length) {
                frameNo -= is.length;
                pts += sampleDurations[curFrag] == null ? avgDur[curFrag] * is.length
                        : ArrayUtil.sumInt(sampleDurations[curFrag]);
                curFrag++;
                globalFrame += is.length;
            } else {
                pts += sampleDurations[curFrag] == null ? avgDur[curFrag] * frameNo
                        : ArrayUtil.sumInt3(sampleDurations[curFrag], 0, (int) frameNo);
                this.curFrag = curFrag;
                this.curFrame = (int) frameNo;
                this.globalFrame = globalFrame + frameNo;
                this.pts = pts;
                adjustOff();

                return true;
            }
        }
        return false;
    }

    private void adjustOff() {
        this.offInChunk = 0;
        for (int i = 0; i < curFrame; i++) {
            this.offInChunk += sizes[curFrag][i];
        }
    }

    @Override
    public long getCurFrame() {
        return globalFrame;
    }

    @Override
    public void seek(double second) throws IOException {
        int curFrag = 0;
        int globalFrame = 0;
        int pts = 0;
        for (int[] is : sizes) {
            int fragDur = sampleDurations[curFrag] != null ? ArrayUtil.sumInt(sampleDurations[curFrag])
                    : avgDur[curFrag] * is.length;
            if (second > fragDur) {
                second -= fragDur;
                pts += fragDur;
                curFrag++;
                globalFrame += is.length;
            } else {
                this.curFrag = curFrag;
                if (sampleDurations[curFrag] != null) {
                    for (curFrame = 0; curFrame < sampleDurations[curFrag].length; curFrame++) {
                        if (second < sampleDurations[curFrag][curFrame])
                            break;
                        second -= sampleDurations[curFrag][curFrame];
                    }
                } else {
                    this.curFrame = (int) (second / avgDur[this.curFrag]);
                    adjustOff();
                }
                pts += sampleDurations[curFrag] == null ? avgDur[curFrag] * this.curFrame
                        : ArrayUtil.sumInt3(sampleDurations[curFrag], 0, this.curFrame);
                this.globalFrame = globalFrame + this.curFrame;
            }
        }
    }

    @Override
    public boolean gotoSyncFrame(long frameNo) throws IOException {
        int curFrag = 0;
        int globalFrame = 0;
        int pts = 0;
        for (int[] is : sizes) {
            if (frameNo > is.length) {
                frameNo -= is.length;
                pts += sampleDurations[curFrag] == null ? avgDur[curFrag] * is.length
                        : ArrayUtil.sumInt(sampleDurations[curFrag]);
                curFrag++;
                globalFrame += is.length;
            } else {
                this.curFrag = curFrag;
                this.curFrame = 0;
                this.globalFrame = globalFrame;
                this.pts = pts;
                offInChunk = 0;
                return true;
            }
        }
        return false;
    }

    public MP4TrackType getTrackType() {
        return TrakBox.getTrackType(trak);
    }

    @Override
    public void close() throws IOException {
        IOException ex = null;
        for (SeekableByteChannel channel : inputs) {
            try {
                channel.close();
            } catch (IOException e) {
                ex = e;
            }
        }
        if (ex != null)
            throw ex;
    }

    public int getNo() {
        return trak.getTrackHeader().getTrackId();
    }

    public void setDurationHint(double arg) {
        this.durationHint = arg;
    }
}

File: src/main/java/org/jcodec/containers/mp4/demuxer/DashMP4DemuxerTrack.kt
package org.jcodec.containers.mp4.demuxer

import org.jcodec.codecs.aac.AACUtils
import org.jcodec.codecs.h264.H264Utils
import org.jcodec.codecs.h264.io.model.SeqParameterSet
import org.jcodec.common.*
import org.jcodec.common.io.NIOUtils
import org.jcodec.common.io.SeekableByteChannel
import org.jcodec.common.model.ColorSpace
import org.jcodec.common.model.Packet.FrameType
import org.jcodec.containers.mp4.MP4Packet
import org.jcodec.containers.mp4.MP4TrackType
import org.jcodec.containers.mp4.MP4Util
import org.jcodec.containers.mp4.boxes.*
import org.jcodec.containers.mp4.boxes.NodeBox.Companion.findAllPath
import org.jcodec.containers.mp4.boxes.NodeBox.Companion.findFirst
import org.jcodec.containers.mp4.boxes.TrakBox.Companion.getTrackType
import org.jcodec.containers.mp4.demuxer.MP4DemuxerTrackMeta.Companion.getCodecPrivateOpaque
import java.io.Closeable
import java.io.File
import java.io.IOException
import java.nio.ByteBuffer
import java.util.*

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * Demuxes one track out of multiple DASH fragments
 *
 * @author The JCodec project
 */
class DashMP4DemuxerTrack(mov: MovieBox?, trak: TrakBox, fragments: Array<Fragment>) : SeekableDemuxerTrack, Closeable {
    private val sizes: Array<IntArray?>
    private val compOffsets: Array<IntArray?>
    private val chunkOffsets: LongArray
    private val avgDur: IntArray
    private val sampleDurations: Array<IntArray?>
    private var offInChunk: Long = 0
    private val inputs: Array<SeekableByteChannel?>
    private var curFrame = 0
    private var curFrag = 0
    private var globalFrame: Long = 0
    private var pts: Long = 0
    private var frameCount = 0
    private var totalDuration: Long = 0
    private val trak: TrakBox
    private val sampleEntries: Array<SampleEntry?>
    private var durationHint = 0.0

    class Fragment(var frag: TrackFragmentBox, var offset: Long, var input: SeekableByteChannel)

    class FragmentComparator : Comparator<Fragment> {
        override fun compare(arg0: Fragment, arg1: Fragment): Int {
            val a = if (arg0.frag.tfdt == null) 0 else arg0.frag.tfdt.getBaseMediaDecodeTime()
            val b = if (arg1.frag.tfdt == null) 0 else arg1.frag.tfdt.getBaseMediaDecodeTime()
            return if (a < b) -1 else if (a == b) 0 else 1
        }
    }

    private fun sortable(fragments: Array<Fragment>): Boolean {
        for (fragment in fragments) {
            if (fragment.frag.tfdt == null) return false
        }
        return true
    }

    @Synchronized
    @Throws(IOException::class)
    override fun nextFrame(): MP4Packet? {
        if (curFrag >= sizes.size) return null
        if (curFrame >= sizes[curFrag]!!.size) {
            curFrag++
            curFrame = 0
            offInChunk = 0
        }
        if (curFrag >= sizes.size) return null
        val size = sizes[curFrag]!![curFrame]
        return getNextFrame(ByteBuffer.allocate(size))!!
    }

    @Throws(IOException::class)
    protected fun readPacketData(input: SeekableByteChannel?, buffer: ByteBuffer?, offset: Long, size: Int): ByteBuffer {
        val result = buffer!!.duplicate()
        synchronized(input!!) {
            input.setPosition(offset)
            NIOUtils.readL(input, result, size)
        }
        result.flip()
        return result
    }

    @Synchronized
    @Throws(IOException::class)
    fun getNextFrame(storage: ByteBuffer?): MP4Packet? {
        if (curFrag >= sizes.size) return null
        if (curFrame >= sizes[curFrag]!!.size) return null
        val size = sizes[curFrag]!![curFrame]
        require(!(storage != null && storage.remaining() < size)) { "Buffer size is not enough to fit a packet" }
        val pktPos = chunkOffsets[curFrag] + offInChunk
        val result = readPacketData(inputs[curFrag], storage, pktPos, size)
        if (result != null && result.remaining() < size) return null
        val hintedDuration = (durationHint * trak.timescale / frameCount).toInt()
        val duration = if (sampleDurations[curFrag] == null) if (avgDur[curFrag] == 0) hintedDuration else avgDur[curFrag] else sampleDurations[curFrag]!![curFrame]
        val sync = curFrame == 0
        var realPts = pts
        if (compOffsets[curFrag] != null) {
            realPts = pts + compOffsets[curFrag]!![curFrame]
        }
        val ftype = if (sync) FrameType.KEY else FrameType.INTER
        val pkt = MP4Packet(result, realPts, trak.timescale, duration.toLong(), globalFrame, ftype, null, 0,
                realPts, 1, pktPos, size, false)
        offInChunk += size.toLong()
        pts += duration.toLong()
        curFrame++
        globalFrame++
        return pkt
    }

    override fun getMeta(): DemuxerTrackMeta {
        val type = getTrackType(trak)
        val t = if (type == MP4TrackType.VIDEO) TrackType.VIDEO else if (type == MP4TrackType.SOUND) TrackType.AUDIO else TrackType.OTHER
        val seekFrames = IntArray(sizes.size)
        var i = 0
        var numFrames = 0
        while (i < sizes.size) {
            seekFrames[i] = numFrames
            numFrames += sizes[i]!!.size
            i++
        }
        val codecPrivate = codecPrivate
        val videoCodecMeta = videoCodecMeta
        val audioCodecMeta = audioCodecMeta
        val codec = Codec.codecByFourcc(sampleEntries[0]!!.fourcc)
        val opaque = getCodecPrivateOpaque(codec,
                sampleEntries[0])
        return MP4DemuxerTrackMeta(t, codec, totalDuration.toDouble(), seekFrames,
                frameCount, codecPrivate, videoCodecMeta, audioCodecMeta, sampleEntries, opaque)
    }

    protected val colorInfo: ColorSpace?
        protected get() {
            val codec = Codec.codecByFourcc(trak.fourcc)
            if (codec == Codec.H264) {
                val avcC = H264Utils.parseAVCC(sampleEntries[0] as VideoSampleEntry?)
                val spsList = avcC.getSpsList()
                if (spsList.size > 0) {
                    val sps = SeqParameterSet.read(spsList[0].duplicate())
                    return sps.chromaFormatIdc
                }
            }
            return null
        }

    private val audioCodecMeta: AudioCodecMeta?
        private get() {
            val type = getTrackType(trak)
            var audioCodecMeta: AudioCodecMeta? = null
            if (type == MP4TrackType.SOUND) {
                val ase = sampleEntries[0] as AudioSampleEntry?
                audioCodecMeta = AudioCodecMeta.fromAudioFormat(ase!!.format)
            }
            return audioCodecMeta
        }

    private val videoCodecMeta: VideoCodecMeta?
        private get() {
            val type = getTrackType(trak)
            var videoCodecMeta: VideoCodecMeta? = null
            if (type == MP4TrackType.VIDEO) {
                videoCodecMeta = VideoCodecMeta.createSimpleVideoCodecMeta(trak.codedSize, colorInfo)
                val pasp = findFirst(sampleEntries[0], "pasp") as PixelAspectExt?
                if (pasp != null) videoCodecMeta.pixelAspectRatio = pasp.rational
            }
            return videoCodecMeta
        }

    // This codec does not have private section
    val codecPrivate: ByteBuffer?
        get() {
            val codec = Codec.codecByFourcc(sampleEntries[0]!!.fourcc)
            if (codec == Codec.H264) {
                val avcC = H264Utils.parseAVCC(sampleEntries[0] as VideoSampleEntry?)
                return H264Utils.avcCToAnnexB(avcC)
            } else if (codec == Codec.AAC) {
                return AACUtils.getCodecPrivate(sampleEntries[0])
            }
            // This codec does not have private section
            return null
        }

    @Throws(IOException::class)
    override fun gotoFrame(frameNo: Long): Boolean {
        var frameNo = frameNo
        var curFrag = 0
        var globalFrame = 0
        var pts = 0L
        for (`is` in sizes) {
            if (frameNo > `is`!!.size) {
                frameNo -= `is`.size.toLong()
                pts += if (sampleDurations[curFrag] == null) avgDur[curFrag] * `is`.size else ArrayUtil.sumInt(sampleDurations[curFrag])
                curFrag++
                globalFrame += `is`.size
            } else {
                pts += if (sampleDurations[curFrag] == null) avgDur[curFrag] * frameNo else ArrayUtil.sumInt3(sampleDurations[curFrag], 0, frameNo.toInt()).toLong()
                this.curFrag = curFrag
                curFrame = frameNo.toInt()
                this.globalFrame = globalFrame + frameNo
                this.pts = pts.toLong()
                adjustOff()
                return true
            }
        }
        return false
    }

    private fun adjustOff() {
        offInChunk = 0
        for (i in 0 until curFrame) {
            offInChunk += sizes[curFrag]!![i]
        }
    }

    override fun getCurFrame(): Long {
        return globalFrame
    }

    @Throws(IOException::class)
    override fun seek(second: Double) {
        var second = second
        var curFrag = 0
        var globalFrame = 0
        var pts = 0
        for (`is` in sizes) {
            val fragDur = if (sampleDurations[curFrag] != null) ArrayUtil.sumInt(sampleDurations[curFrag]) else avgDur[curFrag] * `is`!!.size
            if (second > fragDur) {
                second -= fragDur.toDouble()
                pts += fragDur
                curFrag++
                globalFrame += `is`!!.size
            } else {
                this.curFrag = curFrag
                if (sampleDurations[curFrag] != null) {
                    curFrame = 0
                    while (curFrame < sampleDurations[curFrag]!!.size) {
                        if (second < sampleDurations[curFrag]!![curFrame]) break
                        second -= sampleDurations[curFrag]!![curFrame]
                        curFrame++
                    }
                } else {
                    curFrame = (second / avgDur[this.curFrag]).toInt()
                    adjustOff()
                }
                pts += if (sampleDurations[curFrag] == null) avgDur[curFrag] * curFrame else ArrayUtil.sumInt3(sampleDurations[curFrag], 0, curFrame)
                this.globalFrame = globalFrame + curFrame.toLong()
            }
        }
    }

    @Throws(IOException::class)
    override fun gotoSyncFrame(frameNo: Long): Boolean {
        var frameNo = frameNo
        var curFrag = 0
        var globalFrame = 0
        var pts = 0
        for (`is` in sizes) {
            if (frameNo > `is`!!.size) {
                frameNo -= `is`.size.toLong()
                pts += if (sampleDurations[curFrag] == null) avgDur[curFrag] * `is`.size else ArrayUtil.sumInt(sampleDurations[curFrag])
                curFrag++
                globalFrame += `is`.size
            } else {
                this.curFrag = curFrag
                curFrame = 0
                this.globalFrame = globalFrame.toLong()
                this.pts = pts.toLong()
                offInChunk = 0
                return true
            }
        }
        return false
    }

    val trackType: MP4TrackType?
        get() = getTrackType(trak)

    @Throws(IOException::class)
    override fun close() {
        var ex: IOException? = null
        for (channel in inputs) {
            try {
                channel!!.close()
            } catch (e: IOException) {
                ex = e
            }
        }
        if (ex != null) throw ex
    }

    val no: Int
        get() = trak.trackHeader.trackId

    fun setDurationHint(arg: Double) {
        durationHint = arg
    }

    companion object {
        @JvmStatic
        @Throws(IOException::class)
        fun createFromFiles(files: List<File?>): DashMP4DemuxerTrack {
            val fragments: MutableList<Fragment> = ArrayList()
            var moov: MovieBox? = null
            for (file in files) {
                val channel: SeekableByteChannel = NIOUtils.readableChannel(file)
                for (atom in MP4Util.getRootAtoms(channel)) {
                    if ("moov" == atom.header.fourcc) {
                        moov = atom.parseBox(channel) as MovieBox
                    } else if ("moof".equals(atom.header.fourcc, ignoreCase = true)) {
                        val mfra = atom.parseBox(channel) as MovieFragmentBox
                        val tfra = mfra.tracks[0]
                        fragments.add(Fragment(tfra, atom.offset, channel))
                    }
                }
            }
            return DashMP4DemuxerTrack(moov, moov!!.tracks[0], fragments.toTypedArray())
        }
    }

    init {
        var prevOffset: Long = 0
        var prevSampleCount = 0
        sizes = arrayOfNulls(fragments.size)
        compOffsets = arrayOfNulls(fragments.size)
        chunkOffsets = LongArray(fragments.size)
        avgDur = IntArray(fragments.size)
        sampleDurations = arrayOfNulls(fragments.size)
        this.trak = trak
        inputs = arrayOfNulls(fragments.size)
        sampleEntries = findAllPath(trak, arrayOf("mdia", "minf", "stbl", "stsd", null)) as Array<SampleEntry?>
        if (sortable(fragments)) Arrays.sort(fragments, 0, fragments.size, FragmentComparator())
        var i = 0
        for (fragment in fragments) {
            val frag = fragment.frag
            val trun = frag.trun
            val tfdt = frag.tfdt
            if (tfdt != null) {
                if (i > 0) {
                    avgDur[i - 1] = ((tfdt.getBaseMediaDecodeTime() - prevOffset) / prevSampleCount).toInt()
                    totalDuration = tfdt.getBaseMediaDecodeTime()
                }
                prevOffset = tfdt.getBaseMediaDecodeTime()
            }
            prevSampleCount = trun.getSampleCount().toInt()
            sizes[i] = trun.sampleSizes
            compOffsets[i] = trun.sampleCompositionOffsets
            chunkOffsets[i] = fragment.offset + frag.tfhd.baseDataOffset + frag.trun.getDataOffset()
            if (trun.isSampleDurationAvailable) {
                sampleDurations[i] = trun.sampleDurations
                totalDuration += ArrayUtil.sumInt(sampleDurations[i]).toLong()
            }
            frameCount += sizes[i]!!.size
            inputs[i] = fragment.input
            i++
        }
        if (avgDur.size > 1) {
            avgDur[avgDur.size - 1] = avgDur[avgDur.size - 2]
            totalDuration += avgDur[avgDur.size - 1] * prevSampleCount.toLong()
        }
    }
}